src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -243,22 +306,22 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
     ) -> Result<(Res, Option<String>), ErrorKind<'path>> {
         let cx = self.cx;
 
-        let result = cx.enter_resolver(|resolver| {
-            resolver.resolve_str_path_error(DUMMY_SP, &path_str, ns, module_id)
-        });
-        debug!("{} resolved to {:?} in namespace {:?}", path_str, result, ns);
-        let result = match result {
-            Ok((_, Res::Err)) => Err(()),
-            x => x,
-        };
-
-        if let Ok((_, res)) = result {
-            let res = res.map_id(|_| panic!("unexpected node_id"));
-            // In case this is a trait item, skip the
-            // early return and try looking for the trait.
-            let value = match res {
-                Res::Def(DefKind::AssocFn | DefKind::AssocConst, _) => true,
-                Res::Def(DefKind::AssocTy, _) => false,
+        if let Some(res) = self.resolve_path(path_str, ns, module_id) {
+            match res {
+                Res::Def(DefKind::AssocFn | DefKind::AssocConst, _) => {
+                    if ns != ValueNS {
+                        return Err(ResolutionFailure::WrongNamespace(res, ns).into());
+                    }
```

Comment:
On second thought, I don't think this should ever happen: `resolve_path` delegates to `resolve_str_path_error`, which takes the namespace into account. Can you change this to an assert? I don't expect it to ever fail.
```suggestion
                    assert_eq!(ns, ValueNS);
```